src/features/weak-references.md

```diff
@@ -54,104 +54,133 @@ A JavaScript `WeakMap` is not really _weak_: it actually refers _strongly_ to it
+
+Until now, the solution is to manually unregister the event listener via `dispose` method.
 
 ```js
-const cache = new Map();
+class MovingAvg {
+  constructor(socket) {
+    this.events = [];
+    this.socket = socket;
+    this.listener = (ev) => { this.events.push(ev); };
+    socket.addEventListener('message', this.listener);
+  }
 
-function getImageCached(name) {
-  const ref = cache.get(name);
-  if (ref !== undefined) {
-    const deref = ref.deref();
-    if (deref !== undefined) return deref;
+  dispose() {
+    this.socket.removeEventListener('message', this.listener);
   }
-  const image = performExpensiveOperation(name);
-  const wr = new WeakRef(image);
-  cache.set(name, wr);
-  return image;
+
+  // ...
 }
 ```
 
-But there’s still a problem here: the `Map` still holds on to the `name` strings forever, because those are the keys in the cache. Ideally, those strings would be removed too. The `WeakRef` proposal has a solution for this as well! With the new `FinalizationRegistry` API, we can register a callback to run when the garbage collector zaps a registered object. Such callbacks are known as _finalizers_.
+The downside to this approach is that it is manual memory management. `MovingAvgComponent`, and all other users of the `MovingAvg` class, must remember to call `dispose` or suffer memory leaks. Because the application behavior doesn't depend on the event listener of this diagnostic class, and that the listener is expensive in terms of memory use but not in computation, we really want the listener's lifetime to be logically tied to the `MovingAvg` instance. If its lifetime were tied to the `MovingAvg` instance, then `MovingAvg` could be used like any other JavaScript object whose memory is automatically reclaimed by the garbage collector.
 
-:::note
-**Note:** The finalization callback does not run immediately after garbage-collecting the image object. It either runs at some point in the future, or not at all — the spec doesn’t guarantee that it runs! Keep this in mind when writing code.
-:::
-
-Here, we register a callback to remove keys from the cache when the image objects are garbage-collected:
+`WeakRef`s make it possible to solve the dilemma by creating a _weak reference_ to the actual event listener, and then wrapping that `WeakRef` in an outer event listener. This way, the garbage collector can clean up the actual event listener and the memory that it holds alive, like the `events` array.
 
 ```js
-const cache = new Map();
-
-const finalizationRegistry = new FinalizationRegistry((name) => {
-  const ref = cache.get(name);
-  if (ref !== undefined && ref.deref() === undefined) {
-    cache.delete(name);
+class MovingAvg {
+  constructor(socket) {
+    this.events = [];
+    const listener = (ev) => { this.events.push(ev); }
```

Comment:
```suggestion
    const listener = (ev) => { this.events.push(ev); };
```

Reply:
For reference, we try to follow [the Google JS style guide](https://google.github.io/styleguide/jsguide.html) for any public-facing examples in blog posts and slides, which means: use semicolons, use single quotes instead of double quotes for string literals, prefer `const` over `let`, etc.

Reply:
I'd remove the braces altogether since the return value doesn't matter:

```suggestion
    const listener = (ev) => this.events.push(ev);
```

Reply:
Interestingly, for the exact same reason I would keep the braces :) (Only `return` something non-`undefined` when you need it.)

Reply:
Heh, usually when the caller expects `undefined`, I just return any value.

I know what the ultimate solution would be 😀
```suggestion
    const listener = (ev) => void this.events.push(ev);
```

Reply:
Added semicolon.